[Builtins] Type scheme inference for polymorphic built-in functions

effectfully opened the pull request:
I got curious if there was any way to infer type schemes for polymorphic built-in functions with providing only those types that help resolve inherent ambiguity. I.e. no types at all for functions like `IfThenElse`, `Id` and `Const` and minimal types for functions with higher-kinded type variables. This PR shows how to do that. Note that the shiny new automatic inference machinery does not break the "I'll specify the type signature manually" case.

Supporting the higher-kinds case costs a lot in terms of code complexty and we don't seem to need that at this point and due to stupid behaviour of GHC I can't handle the general case anyway, so I'm going to rewrite this to handle only type variables of kind `*`. But the temptation to show off was too high, so I couldn't resist showing this development in its current form.

effectfully commented on plutus-core/test/Evaluation/DynamicBuiltins/Definition.hs:
```diff
@@ -131,35 +131,27 @@ defBuiltinsRuntimeExt = toBuiltinsRuntime (defDefaultFunDyn, ()) (defaultCostMod
-                   , b ~ Opaque term (TyVarRep ('TyNameRep "b" 1))
-                   )
-                => a -> b -> a)
+            const
```
Isn't

effectfully commented on plutus-core/test/Evaluation/DynamicBuiltins/Definition.hs:
```diff
@@ -131,35 +131,27 @@ defBuiltinsRuntimeExt = toBuiltinsRuntime (defDefaultFunDyn, ()) (defaultCostMod
     toBuiltinMeaning Id =
         toStaticBuiltinMeaning
-            (Prelude.id :: a ~ Opaque term (TyVarRep ('TyNameRep "a" 0)) => a -> a)
+            Prelude.id
```
that

effectfully commented on plutus-core/test/Evaluation/DynamicBuiltins/Definition.hs:
```diff
@@ -131,35 +131,27 @@ defBuiltinsRuntimeExt = toBuiltinsRuntime (defDefaultFunDyn, ()) (defaultCostMod
-            (Prelude.id
-                :: a ~ Opaque term (TyAppRep (TyVarRep ('TyNameRep "f" 0)) (Transparent Integer))
-                => a -> a)
+            (Prelude.id :: a ~ f Integer => a -> a)
```
nice

effectfully commented on plutus-core/test/Evaluation/DynamicBuiltins/Definition.hs:
```diff
@@ -131,35 +131,27 @@ defBuiltinsRuntimeExt = toBuiltinsRuntime (defDefaultFunDyn, ()) (defaultCostMod
-            (Prelude.id
-                :: a ~ Opaque term (ListRep (Opaque term (TyVarRep ('TyNameRep "a" 0))))
-                => a -> a)
+            (Prelude.id :: la ~ Opaque term (ListRep a) => la -> la)
```
?

effectfully commented:
Removed support for higher kinds. I'll write some docs once I'm done with at least an initial version of HOAS.

kwxm commented:
Not as bad as I'd feared, although I still don't understand it! It needs surprisingly little extra code ...

kwxm commented:
I wonder if you could write a blog entry or something about this stuff.  I think @michaelpj said a while ago that the company was encouraging people to blog about technical things that might be of wider interest.

effectfully commented:
> I wonder if you could write a blog entry or something about this stuff.

That's exactly what I'm going to do.

> I think @michaelpj said a while ago that the company was encouraging people to blog about technical things that might be of wider interest.

I tried to do that a few times in the past, but it did not work out. I'll just post something in my personal blog.

effectfully commented:
Explained the ideas behind this PR in these two posts: [1](https://github.com/effectfully/sketches/tree/master/poly-type-of-saga/part1-try-unify). [2](https://github.com/effectfully/sketches/tree/master/poly-type-of-saga/part2-enumerate-type-vars).

kwxm commented:
That see

> Explained the ideas behind this PR in these two posts: [1](https://github.com/effectfully/sketches/tree/master/poly-type-of-saga/part1-try-unify), [2](https://github.com/effectfully/sketches/tree/master/poly-type-of-saga/part2-enumerate-type-vars).

Those seem pretty clear.  I still don't understand it, but it's good to have an explanation written down for future reference!

A couple of typos: post 1 contains the word "firt", and post 2 says "the base case if a bit more interesting".

I'm wondering what you're going to come up with next.  Type-level `goto`?

effectfully commented:
Refactored & added docs, should be ready for review.

@kwxm 

> A couple of typos: post 1 contains the word "firt", and post 2 says "the base case if a bit more interesting".

Thanks, fixed.

> I'm wondering what you're going to come up with next. Type-level goto?

Hm, interesting... :)

michaelpj commented:
Doesn't compile?

effectfully commented:
> Doesn't compile?

I was sure I fixed that. Anyway, just fixed it by making the machinery a bit less powerful, going to write some docs and tests and then push.

effectfully commented:
It's a GHC bug, by the way. The error is

```
    • Occurs check: cannot construct the infinite type:
        p0 ~ FoldArgs (GetArgs p0) res0
```

That's not an infinite type. The LHS is actually equal to the RHS: `GetArgs` extracts all argument types and `FoldArgs` folds them back into the original type. We need this type gymnastics in order to be able to write `a` as the type of the Haskell denotation argument, so that it can be inferred, while at the same time retaining the ability to fit this type into 

```haskell
data BuiltinMeaning term dyn cost =
    forall args res. BuiltinMeaning
        (TypeScheme term args res)
        (dyn -> FoldArgs args res)
        (cost -> FoldArgsEx args)
```

where `a` is spelled as `FoldArgs args res` (we can't just use `a` there, because `TypeScheme` uses extracted `args` and so the denotation also has to be defined in terms of them for types to match).

So we only get this error because occurs check is broken in GHC.

effectfully commented:
Fixed the problem, now ready for review.
